### PR TITLE
Adds support of custom response handler(s)

### PR DESCRIPTION
**Pull request by jo-asakura**
You can now provide custom response handler(s) using `configureFetch` call:
```
configureFetch({
  headers: {
    'X-CSRF-Token': window.env.csrfToken,
  },
  responseMiddleware: [
    function nonAuthHandler (response, next) {
      if (response && response.status === 401) {
        return window.location = 'http://localhost:3000/authenticate/users/sign_in';
      }
      return next(response);
    },
    function anotherHandler (response, next) {
      response.hello = 'world';
      return next(response);
    }
  ]
});
```

`responseMiddleware` can be either a function or an array of functions that will be executed in sequence.

In defined handler(s) you get access to _raw_ response object before any `react-data-action` modification is done.

Demo:
kapture 2017-07-05 at 16 03 24

### CODE REVIEW

**Comment by justinseiter** [comment]
@jo-asakura What are your thoughts on making `runResponseHandlers` part of the promise chain? I'm thinking about race conditions; for example, what if our handler was responding to a `204` (a code we address later in our logic)? Especially considering we could pass an array of functions.

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "react-data-actions",
-  "version": "0.7.4",
+  "version": "0.7.5",
```
Comment (mjswensen) [inline package.json]: I'd argue that this would be a minor version change (`0.8.0`).

File: src/configureFetch.js [hunk 2]
```diff
@@ -6,12 +8,34 @@ const requestConfiguration = {
   },
 };
 
-function getFetchConfiguration () {
-  return requestConfiguration;
-}
+const responseHandlers = {
+  handlers: [],
+  set: (handlers) => {
+    responseHandlers.handlers = Array.isArray(handlers) ? handlers : [handlers];
```
Comment (mjswensen) [inline src/configureFetch.js]: I do love the flexibility of either passing a function or an array of functions. However, the name `responseHandlers` seems like a misnomer for a single function. You could take the webpack config approach and provide both a `responseHandler` (for a single function) and a `responseHandlers` (for an array) in the API. Alternatively, you could use a more generic name that works for both, like `interceptResponse`, `responseMiddleware`, or `withResponse`.

File: src/configureFetch.js [hunk 3]
```diff
@@ -6,12 +8,34 @@ const requestConfiguration = {
...
+  },
+  run: (response) => {
```
Comment (mjswensen) [inline src/configureFetch.js]: I definitely agree with @justinseiter that this should be plugged into the promise chain to make it nice and uniform. `_.identity` could then be a nice default for the option if none is provided.

The other advantage to this is that people could use it for _transforming_ a response in addition to performing side effects like a redirect.

**Comment by mrdeadsake** [comment]
So in your example above, where does this extra code go? Like every single action would need to be reconfigured to include this?

File: src/configureFetch.js [hunk 4]
```diff
@@ -6,12 +8,41 @@ const requestConfiguration = {
...
+      return Promise.resolve(response);
+    }
+
+    return responseMiddleware.handlers.reduce((aggr, handler) => {
+      return aggr.then((r) => {
+        return new Promise((resolve) => {
+          return handler(r, resolve);
+        });
+      });
+    }, Promise.resolve(response));
```
Comment (justinseiter) [inline src/configureFetch.js]: Wondering if `Promise.all()` would be a better fit here as we are wrapping each function in a promise?

**Comment by justinseiter** [comment]
Awesome Alex! 👍 
One last Q, what do you think about adding a `catch` for the `responseMiddleware` in case something goes wrong with someone's custom functions.

File: src/configureFetch.js (hunk 4 above)
Comment (jo-asakura) [inline src/configureFetch.js]: `Promise.all` resolves all promises in random order and we want middleware-like way when we can run promises in sequence and pass modified response from one promise to another.

**Comment by jo-asakura** [comment]
@justinseiter The problem is even if we catch that error the promises chain will break and `react-data-actions` will fail... promises are that bad 😄 You get same behavior in express or koa, if your middleware fail your entire app fails.

**Comment by jo-asakura** [comment]
@mrdeadsake It goes in `application.js` where you call `configureFetch` for `react-data-actions`: https://github.com/nuvi/gavin/blob/master/client/application.js#L30

**Pull request merged**